### ISSUE TITLE
Abnormal time consumption of  non_max_suppression() 

### ISSUE THREAD

**Issue by Zz928** [ultralytics/yolov5#12183]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### YOLOv5 Component

Detection

### Bug

Abnormal time consumption of  non_max_suppression()
detect one second apart,  non_max_suppression() It takes a long time
if dectect continuous,no time interval,non_max_suppression().Time consuming is very short.

### Environment

YOLOv5  2023-9-30 Python-3.10.11 torch-2.0.1+cu118 CUDA:0 (NVIDIA GeForce GTX 1660 Ti, 6144MiB)

### Minimal Reproducible Example

import argparse
import csv
import os
import platform
import sys
from pathlib import Path
import numpy as np
import time
import torch
import threading
from time import perf_counter as tp

FILE = Path(__file__).resolve()
ROOT = FILE.parents[0]  # YOLOv5 root directory
if str(ROOT) not in sys.path:
    sys.path.append(str(ROOT))  # add ROOT to PATH
ROOT = Path(os.path.relpath(ROOT, Path.cwd()))  # relative

from ultralytics.utils.plotting import Annotator, colors, save_one_box

from yolov5.utils.augmentations import letterbox
from yolov5.models.common import DetectMultiBackend
from yolov5.utils.dataloaders import IMG_FORMATS, VID_FORMATS, LoadImages, LoadScreenshots, LoadStreams
from yolov5.utils.general import (LOGGER, Profile, check_file, check_img_size, check_imshow, check_requirements, colorstr, cv2,
                           increment_path, non_max_suppression, print_args, scale_boxes, strip_optimizer, xyxy2xywh)
from yolov5.utils.torch_utils import select_device, smart_inference_mode

def find_test(
    source,conf_thres=0.25,iou_thres=0.45,max_det=1000, classes:int=None, agnostic_nms=False, augment=False):
    weights = r"yolo/pretrained/yolov5s.pt"
    device =torch.device("cuda:0")
    dnn=False
    data = False
    half=False
    imgsz=(640,640)
    model = DetectMultiBackend(weights, device=device, dnn=dnn, data=data, fp16=half)
    
    for i in range(100):
        start=tp()
        im0 = cv2.imread(source)
            # contiguous        
        im = letterbox(im0,imgsz, stride=32, auto=True)[0]  # padded resize
        im = im.transpose((2, 0, 1))[::-1]  # HWC to CHW, BGR to RGB
        im = np.ascontiguousarray(im)  # contiguous

        im = torch.from_numpy(im).to(model.device)
        im = im.half() if model.fp16 else im.float()  # uint8 to fp16/32
        im /= 255  # 0 - 255 to 0.0 - 1.0
        if len(im.shape) == 3:
            im = im[None]  # expand for batch dim
        pred = model(im, augment=augment, visualize=False)
        pred = non_max_suppression(pred, conf_thres, iou_thres, classes, agnostic_nms, max_det=max_det)
        for i, det in enumerate(pred):  # per image

            gn = torch.tensor(im0.shape)[[1, 0, 1, 0]]  # normalization gain whwh
            if len(det):
                # Rescale boxes from img_size to im0 size
                det[:, :4] = scale_boxes(im.shape[2:], det[:, :4], im0.shape).round()

                for *xyxy, conf, cls in reversed(det):

                    xywh = (xyxy2xywh(torch.tensor(xyxy).view(1, 4)) / gn).view(-1).tolist()  # normalized xywh
                    
                print(round(((tp()-start)*1000),6))
                time.sleep(1)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@Zz928 the abnormal time consumption of `non_max_suppression()` that you're experiencing might be due to the continuous running of the detection function without any time interval. 

In your provided code, there is a `time.sleep(1)` at the end of each iteration within the for loop. This introduces a time delay that allows the `non_max_suppression()` function to complete quickly. However, when you remove the time delay, the function takes longer to execute.

`non_max_suppression()` is a computationally intensive operation, and its execution time can be affected by various factors such as the number of detections and the configuration of your setup.

If you want to optimize the performance of `non_max_suppression()`, you can consider the following:

1. Reduce the number of detections by adjusting the confidence threshold (`conf_thres`) and the maximum number of detections (`max_det`).
2. Experiment with different values of the intersection over union threshold (`iou_thres`) to see if it improves the execution time.
3. Ensure that you are using an optimized environment, such as a GPU with CUDA support, to take advantage of hardware acceleration.

Please note that YOLOv5 is a community-driven project, and the Ultralytics team maintains and contributes to it. The execution time of `non_max_suppression()` is influenced by the algorithms and optimizations implemented by the YOLO community.

Let me know if there's anything else I can help you with!

**Comment by Zz928** [comment]
> 用一个非空的文件。 the abnormal time consumption of `non_max_suppression()` that you're experiencing might be due to the continuous running of the detection function without any time interval.
> 
> In your provided code, there is a `time.sleep(1)` at the end of each iteration within the for loop. This introduces a time delay that allows the `non_max_suppression()` function to complete quickly. However, when you remove the time delay, the function takes longer to execute.
> 
> `non_max_suppression()` is a computationally intensive operation, and its execution time can be affected by various factors such as the number of detections and the configuration of your setup.
> 
> If you want to optimize the performance of `non_max_suppression()`, you can consider the following:
> 
> 1. Reduce the number of detections by adjusting the confidence threshold (`conf_thres`) and the maximum number of detections (`max_det`).
> 2. Experiment with different values of the intersection over union threshold (`iou_thres`) to see if it improves the execution time.
> 3. Ensure that you are using an optimized environment, such as a GPU with CUDA support, to take advantage of hardware acceleration.
> 
> Please note that YOLOv5 is a community-driven project, and the Ultralytics team maintains and contributes to it. The execution time of `non_max_suppression()` is influenced by the algorithms and optimizations implemented by the YOLO community.
> 
> Let me know if there's anything else I can help you with!


I mean to say, after conducting an inference，after a time，non_max_suppression() the execution time is more than 100ms，.if after conducting an inference,Start the next inference immediately,no time interval.non_max_suppression() the execution time is less than 5ms. My code use ‘time.sleep(1)’ as a time interval，used to reproduce problems. Under normal circumstances,For the same image,non_max_suppression() . The time consumption should be similar.but now,The shorter the interval after each inference is completed,  the shorter the consumption time of non_max_suppression().for example,use 'time.sleep(1)',non_max_suppression()--100ms;'time.sleep(0.1)',non_max_suppression()--50ms;'time.sleep(0.01)',non_max_suppression()--10ms;  'time.sleep()' Time consuming operations after simulation inference is completed

**Comment by glenn-jocher** [comment]
@Zz928 the abnormal time consumption you're experiencing with `non_max_suppression()` after conducting an inference without any time interval can be explained by the behavior you described. When you remove the time delay, the execution time of `non_max_suppression()` decreases significantly.

This behavior can be attributed to several factors. The time delay introduced by `time.sleep()` allows some processing time for the `non_max_suppression()` function to complete before the next inference starts. This can result in faster execution times for `non_max_suppression()` when compared to immediately starting the next inference without any delay.

The execution time of `non_max_suppression()` can vary depending on the number of detections and the configuration of your setup. By adjusting the confidence threshold (`conf_thres`) and the maximum number of detections (`max_det`), you can reduce the number of detections, which may improve the performance of `non_max_suppression()`.

Additionally, experimenting with different values of the intersection over union threshold (`iou_thres`) may also impact the execution time of `non_max_suppression()`. You can try different values to see if they affect the performance.

It is important to note that YOLOv5 is a community-driven project, and the execution time of `non_max_suppression()` is influenced by the algorithms and optimizations implemented by the YOLO community. Therefore, the behavior you're observing might be a result of the underlying implementation.

If you have any further questions or need additional assistance, please let me know.

**Comment by Zz928** [comment]
Your help was very much appreciated. But i think you didn't understand what I was saying. 
I start detect a picture. Waiting for this detection to complete.Then wait for 1 s. Then I detect a picture again.  The execution time of this detection is 50 ms.
if  I start detect a picture. Waiting for this detection to complete.Then wait for 10 s. Then I detect a picture again. The execution time of this detection is 100 ms.
if  I start detect a picture. Waiting for this detection to complete.Then no wait.Then I detect a picture again. The execution time of this detection is 5 ms.

The execution time of this detection is not normal.  the non_max_suppression()  function is the main time-consuming program.
Time interval between two number of detections should not affect the execution time of detection.I think it's not normal. 

Thank you again for your answer.

**Comment by glenn-jocher** [comment]
@Zz928 i apologize for any misunderstanding in my previous response. Thank you for providing further clarification.

Based on your description, it seems that the time interval between two consecutive detections is affecting the execution time of the detection process. Specifically, when you introduce a longer time delay (e.g., 10 seconds) before starting the next detection, the execution time of `non_max_suppression()` increases to 100 ms. However, when you remove the time delay and immediately start the next detection, the execution time decreases significantly to 5 ms.

While it's expected that the execution time of `non_max_suppression()` can vary depending on factors such as the number of detections and their complexity, the significant difference in execution time due to the time interval is not a typical behavior.

To further investigate this issue and understand the possible reasons behind this behavior, it would be helpful to provide additional information such as the specific versions of YOLOv5 and any modifications you've made to the code. Additionally, specifying the hardware setup and configuration you're using could also provide more context for troubleshooting this issue.

Thank you for bringing this issue to our attention, and we appreciate your patience as we work towards resolving it. If you have any further details or questions, please let us know, and we'll be happy to assist you.

**Comment by Zz928** [comment]
> @Zz928 i apologize for any misunderstanding in my previous response. Thank you for providing further clarification.
> 
> Based on your description, it seems that the time interval between two consecutive detections is affecting the execution time of the detection process. Specifically, when you introduce a longer time delay (e.g., 10 seconds) before starting the next detection, the execution time of `non_max_suppression()` increases to 100 ms. However, when you remove the time delay and immediately start the next detection, the execution time decreases significantly to 5 ms.
> 
> While it's expected that the execution time of `non_max_suppression()` can vary depending on factors such as the number of detections and their complexity, the significant difference in execution time due to the time interval is not a typical behavior.
> 
> To further investigate this issue and understand the possible reasons behind this behavior, it would be helpful to provide additional information such as the specific versions of YOLOv5 and any modifications you've made to the code. Additionally, specifying the hardware setup and configuration you're using could also provide more context for troubleshooting this issue.
> 
> Thank you for bringing this issue to our attention, and we appreciate your patience as we work towards resolving it. If you have any further details or questions, please let us know, and we'll be happy to assist you.

 Thank you for your answer.
yolov5 v7.0  win10  Python-3.10.11  torch-2.0.1+cu118 CUDA:0 (NVIDIA GeForce GTX 1660 Ti, 6144MiB)  

my code:
def find_test(
    source,conf_thres=0.25,iou_thres=0.45,max_det=1000, classes:int=None, agnostic_nms=False, augment=False):
    weights = r"yolo/pretrained/yolov5s.pt"
    device =torch.device("cuda:0")
    dnn=False
    data = False
    half=False
    imgsz=(640,640)
    model = DetectMultiBackend(weights, device=device, dnn=dnn, data=data, fp16=half)
    
    for i in range(100):

        im0 = cv2.imread(source)
            # contiguous        
        im = letterbox(im0,imgsz, stride=32, auto=True)[0]  # padded resize
        im = im.transpose((2, 0, 1))[::-1]  # HWC to CHW, BGR to RGB
        im = np.ascontiguousarray(im)  # contiguous

        im = torch.from_numpy(im).to(model.device)
        im = im.half() if model.fp16 else im.float()  # uint8 to fp16/32
        im /= 255  # 0 - 255 to 0.0 - 1.0
        if len(im.shape) == 3:
            im = im[None]  # expand for batch dim
        pred = model(im, augment=augment, visualize=False)
        pred = non_max_suppression(pred, conf_thres, iou_thres, classes, agnostic_nms, max_det=max_det)
        for i, det in enumerate(pred):  # per image

            gn = torch.tensor(im0.shape)[[1, 0, 1, 0]]  # normalization gain whwh
            if len(det):
                # Rescale boxes from img_size to im0 size
                det[:, :4] = scale_boxes(im.shape[2:], det[:, :4], im0.shape).round()

                for *xyxy, conf, cls in reversed(det):

                    xywh = (xyxy2xywh(torch.tensor(xyxy).view(1, 4)) / gn).view(-1).tolist()  # normalized xywh                   

                time.sleep(1)

**Comment by glenn-jocher** [comment]
@Zz928 thank you for providing the additional information. Based on the code snippet you shared, I can see that you are using YOLOv5 version 7.0 on Windows 10 with Python 3.10.11 and Torch 2.0.1+cu118. Your specific hardware setup includes an NVIDIA GeForce GTX 1660 Ti with 6144MiB of memory.

In your code, you are performing object detection using YOLOv5 on a source image. You're using the `DetectMultiBackend` class to load the weights of the YOLOv5 model and configure the detection parameters such as confidence threshold (`conf_thres`), intersection over union threshold (`iou_thres`), maximum number of detections (`max_det`), classes, agnostic NMS, and augmentation.

You mentioned that the execution time of the `non_max_suppression()` function varies depending on the time interval between two consecutive detections. Specifically, when you introduce a longer time delay (e.g., 10 seconds) before starting the next detection, the execution time of `non_max_suppression()` increases to 100 ms. However, when you remove the time delay and immediately start the next detection, the execution time decreases significantly to 5 ms.

While it's expected that the execution time of `non_max_suppression()` can vary depending on factors such as the number of detections and their complexity, the significant difference in execution time due to the time interval is not a typical behavior.

To further investigate this issue and understand the possible reasons behind this behavior, it would be helpful to know if you have made any modifications to the YOLOv5 code or if you are using the official YOLOv5 repository without any modifications. Additionally, providing details about the specific hardware setup and configuration you are using would also be useful for troubleshooting this issue.

Thank you for bringing this issue to our attention, and we appreciate your patience as we work towards resolving it. If you have any further details or questions, please let us know, and we'll be happy to assist you.
